<a href="https://colab.research.google.com/github/gabiuxo/Algoritmos-de-Aprendizaje-Automatico/blob/main/PracticaViernesTema14Parte2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Práctica viernes - Tema 14 Parte 2: Supervisión, calibración y comunicación

**Gabriel Elizondo Martinez**  
**Matrícula:** AL07009102


## Reto 1. Entrenamiento y generación de probabilidades de referencia


In [1]:
import numpy as np

from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# create the base data and split it
X, y = make_classification(n_samples=600, n_features=5, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# train the classifier and extract continuous probabilities
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)
y_probs = model.predict_proba(X_test)[:, 1]

print("Reference probabilities generated for", len(y_probs), "test samples.")
print(f"Minimum probability: {y_probs.min():.4f}")
print(f"Maximum probability: {y_probs.max():.4f}")
print(f"Average probability: {y_probs.mean():.4f}")


Reference probabilities generated for 120 test samples.
Minimum probability: 0.0003
Maximum probability: 1.0000
Average probability: 0.5125


Entrené el modelo y obtuve una probabilidad para cada una de las 120 observaciones de prueba. Estas probabilidades permiten evaluar no solo la clase final, sino también qué tan confiado está el modelo en cada predicción.


## Reto 2. Evaluación de calibración con Brier Score


In [2]:
# calculate the individual Brier Score for the first test observation
probability = y_probs[0]
actual_label = y_test[0]
individual_brier = (probability - actual_label) ** 2

print(f"Probability: {probability:.4f}")
print(f"Actual label: {actual_label}")
print(f"Individual Brier Score: {individual_brier:.4f}")


Probability: 0.1443
Actual label: 0
Individual Brier Score: 0.0208


Calculé el Brier Score de la primera observación con la fórmula (p - y)². Un resultado cercano a cero significa que la probabilidad reportada fue más coherente con el resultado real, lo cual es importante cuando el umbral de decisión se ajusta con precisión.


## Reto 3. Runbook de incidentes ante alerta de drift


In [3]:
import pandas as pd

# define the runbook for an 18% increase in positive predictions
runbook = pd.DataFrame([
    {"Paso": "Validar", "Acción": "Confirmar que el aumento de 18% no sea un error de datos, cálculo o servicio.", "Responsable": "Operación"},
    {"Paso": "Diagnosticar", "Acción": "Comparar datos y scores entre la ventana base y la reciente para identificar la causa.", "Responsable": "Especialista técnico"},
    {"Paso": "Mitigar", "Acción": "Revisar el umbral y aplicar una medida temporal si el cambio afecta la operación.", "Responsable": "Propietario del modelo"},
    {"Paso": "Documentar", "Acción": "Registrar la alerta, evidencia, decisión tomada y seguimiento en la bitácora.", "Responsable": "Operación"},
])

print("Alert: positive prediction rate increased by 18%")
print(runbook.to_string(index=False))


Alert: positive prediction rate increased by 18%
        Paso                                                                                 Acción            Responsable
     Validar          Confirmar que el aumento de 18% no sea un error de datos, cálculo o servicio.              Operación
Diagnosticar Comparar datos y scores entre la ventana base y la reciente para identificar la causa.   Especialista técnico
     Mitigar      Revisar el umbral y aplicar una medida temporal si el cambio afecta la operación. Propietario del modelo
  Documentar          Registrar la alerta, evidencia, decisión tomada y seguimiento en la bitácora.              Operación


Definí un runbook de cuatro pasos para evitar una respuesta improvisada ante un cambio de 18% en la tasa de positivos. Cada etapa tiene una acción y un responsable para que la alerta pueda revisarse y dejar evidencia de la decisión tomada.


## Reto 4. Resumen ejecutivo de negocio


El modelo ayuda a identificar casos con mayor probabilidad de requerir atención, lo que permite priorizar recursos y actuar antes. Su valor depende de que las señales generadas sean confiables y se revisen continuamente. El principal riesgo es que un cambio en los datos reduzca la utilidad de sus resultados; por ello, mantendría el monitoreo de las predicciones y revisaría cualquier cambio relevante antes de modificar la operación.

## Conclusión

En esta práctica comprobé que un modelo debe supervisarse y comunicarse de forma clara después de entrenarse. Revisé la calidad de sus probabilidades, definí qué hacer ante una alerta y resumí su valor y riesgo para una audiencia de negocio.
